client/allocrunner/identity_hook.go

```diff
@@ -0,0 +1,248 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package allocrunner
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/client/allocrunner/interfaces"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	"github.com/hashicorp/nomad/helper"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// IdentitySigner is the interface needed to retrieve signed identities for
+// workload identities. At runtime it is implemented by *widmgr.WIDMgr.
+type IdentitySigner interface {
+	SignIdentities(minIndex uint64, req []*structs.WorkloadIdentityRequest) ([]*structs.SignedWorkloadIdentity, error)
+}
+
+type identityHook struct {
+	ar            *allocRunner
+	hookResources *cstructs.AllocHookResources
+	widmgr        IdentitySigner
+	logger        log.Logger
+
+	// minWait is the minimum amount of time to wait before renewing. Settable to
+	// ease testing.
+	minWait time.Duration
+
+	stopCtx context.Context
+	stop    context.CancelFunc
+}
+
+func newIdentityHook(ar *allocRunner, hookResources *cstructs.AllocHookResources, logger log.Logger) *identityHook {
+	// Create a context for the renew loop. This context will be canceled when
+	// the allocation is stopped or agent is shutting down
+	stopCtx, stop := context.WithCancel(context.Background())
+
+	h := &identityHook{
+		ar:            ar,
+		hookResources: hookResources,
+		minWait:       10 * time.Second,
+		stopCtx:       stopCtx,
+		stop:          stop,
+	}
+	h.logger = logger.Named(h.Name())
+	return h
+}
+
+func (*identityHook) Name() string {
+	return "identity"
+}
+
+func (h *identityHook) Prerun() error {
+	signedWIDs := map[string]*structs.SignedWorkloadIdentity{}
+
+	// we need to know the amount of tasks to create a buffered
+	// SignedTaskIdentities channel
+	numberOfTasks := len(h.ar.tasks)
+
+	for _, t := range h.ar.tasks {
+		task := t.Task()
+		if task == nil {
+			// hitting this means a bug, but better safe than sorry
+			continue
+		}
+
+		var err error
+		signedWIDs, err = h.getIdentities(h.ar.Alloc(), task)
+		if err != nil {
+			return fmt.Errorf("error fetching alternate identities: %w", err)
+		}
+
+		// publish task identities inside hookResources, so that taskrunner
+		// hooks can also use them.
+		h.hookResources.SignedTaskIdentities = make(chan map[string]*structs.SignedWorkloadIdentity, numberOfTasks)
+		h.hookResources.SignedTaskIdentities <- signedWIDs
```

Comment:
This looks like we're creating a channel for each task, which we definitely need. But aren't we overwriting the `hookResources.SignedTaskIdentities` in the process? That means only the channel for the last task will be available at this location by the time the taskrunner `identity_hooks` run. 

(I'm pretty sure that when the other channels fall out of scope they'll get GC'd even though they're full, unlike a leaked goroutine.)